service_discovery/proto/chariott/v1/chariott_registry.proto

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+// SPDX-License-Identifier: MIT
+
+syntax = "proto3";
+package chariott_registry;
+
+// Service Registry definition
+//
+// The protobuf definitions for Chariott's service registry
+
+// The entry point for the Registry gRPC Service.
+service Registry {
+  // Register, or add a service to the registry
+  rpc Register(RegisterRequest) returns (RegisterResponse) {}
+
+  // Unregister, or remove a service from the registry
+  rpc Unregister(UnregisterRequest) returns (UnregisterResponse) {}
+
+  // Discover, or retrieve the metadata for a single service given its fully qualified name
+  rpc DiscoverService(DiscoverServiceRequest) returns (DiscoverServiceResponse) {}
+
+  // Discover a list of services given their namespace
+  rpc DiscoverByNamespace(DiscoverByNamespaceRequest) returns (DiscoverByNamespaceResponse) {}
+
+  // Inspect, or retrieve all contents of the service registry
+  rpc Inspect(InspectRequest) returns (InspectResponse) {}
+}
+
+// Represenation of a service including all of its metadata that the registry stores
+message ServiceMetadata {
+  string namespace = 1;
+  string name = 2;
+  string version = 3;
+  string uri = 4;
+  string communication_kind = 5;
+  string communication_reference = 6;
+}
+
+// Status of a registration operation
+enum RegistrationStatus {
```

Comment:
what about just failing if trying to register something already registered? if we have requirements about doing update we can add that option in the future, for example through ways of specifying if one wants to overwrite and so on during the Register operation

Reply:
Upon further thought, I agree that we can leave it simple for now, with no option to "update" a registration once it's been registered. This also brings up some questions that we have about the behavior of Chariott in terms of should we allow multiple running instances of the same version of a service? How will we determine uniqueness? etc. These will come up in future design iterations. For simplicity's sake and until we have a scenario for update, I will remove the "update" part of the register.